gateway/handlers/update_handler.go

```diff
@@ -1,16 +1,130 @@
 	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := context.Background()
+
 		defer r.Body.Close()
-		w.WriteHeader(http.StatusNotImplemented)
+		body, _ := ioutil.ReadAll(r.Body)
+
+		request := requests.CreateFunctionRequest{}
+		err := json.Unmarshal(body, &request)
+		if err != nil {
+			log.Println("Error parsing request:", err)
+			w.WriteHeader(http.StatusBadRequest)
+			w.Write([]byte(err.Error()))
+			return
+		}
+
+		serviceInspectopts := types.ServiceInspectOptions{
+			InsertDefaults: true,
+		}
+
+		service, _, err := c.ServiceInspectWithRaw(ctx, request.Service, serviceInspectopts)
+		if err != nil {
+			log.Println("Error inspecting service", err)
+			w.WriteHeader(http.StatusNotFound)
+			w.Write([]byte(err.Error()))
+			return
+		}
+
+		updateSpec(&request, &service.Spec, maxRestarts)
+
+		updateOpts := types.ServiceUpdateOptions{}
+		updateOpts.RegistryAuthFrom = types.RegistryAuthFromSpec
+
+		if len(request.RegistryAuth) > 0 {
+			auth, err := BuildEncodedAuthConfig(request.RegistryAuth, request.Image)
+			if err != nil {
+				log.Println("Error building registry auth configuration:", err)
+				w.WriteHeader(http.StatusBadRequest)
+				w.Write([]byte("Invalid registry auth"))
+				return
+			}
+			updateOpts.EncodedRegistryAuth = auth
+		}
+
+		response, err := c.ServiceUpdate(ctx, service.ID, service.Version, service.Spec, updateOpts)
+		if err != nil {
+			log.Println("Error updating service:", err)
+			w.WriteHeader(http.StatusBadRequest)
+			w.Write([]byte("Update error: " + err.Error()))
+			return
+		}
+		log.Println(response.Warnings)
+	}
+}
+
+func updateSpec(request *requests.CreateFunctionRequest, spec *swarm.ServiceSpec, maxRestarts uint64) {
+
+	linuxOnlyConstraints := []string{"node.platform.os == linux"}
+	restartDelay := time.Second * 5
+
+	constraints := []string{}
+	if request.Constraints != nil && len(request.Constraints) > 0 {
+		constraints = request.Constraints
+	} else {
+		constraints = linuxOnlyConstraints
+	}
+
+	nets := []swarm.NetworkAttachmentConfig{
+		{Target: request.Network},
+	}
+
+	spec.TaskTemplate.RestartPolicy.MaxAttempts = &maxRestarts
+	spec.TaskTemplate.RestartPolicy.Condition = swarm.RestartPolicyConditionAny
+	spec.TaskTemplate.RestartPolicy.Delay = &restartDelay
+	spec.TaskTemplate.ContainerSpec.Image = request.Image
+	spec.TaskTemplate.ContainerSpec.Labels = map[string]string{
+		"function": "true",
+		"uid":      fmt.Sprintf("%d", time.Now().Nanosecond()),
+	}
+	spec.TaskTemplate.Networks = nets
+	spec.TaskTemplate.Placement = &swarm.Placement{
+		Constraints: constraints,
+	}
+
+	spec.Annotations = swarm.Annotations{
+		Name: request.Service,
+	}
+
+	spec.RollbackConfig = &swarm.UpdateConfig{
+		FailureAction: "pause",
+	}
+
+	spec.UpdateConfig = &swarm.UpdateConfig{
+		Parallelism:   1,
+		FailureAction: "rollback",
+	}
+
+	var env []string
+	if len(request.EnvProcess) > 0 {
```

Comment:
is this common code between create / update?

Reply:
yes, there are chunks of repetition, will have a look at moving stuff into a common location